on_device_training/mobilevit/mobilevit_console/mobilevit_console/Program.cs

```diff
@@ -0,0 +1,168 @@
+
+            return dl;
+        }
+
+        public static void CreateAndRunTrainingSession(int numEpochs, int batchSize, DataLoader dl)
+        {
+            string parentDir = Directory.GetParent(Directory.GetCurrentDirectory()).Parent.Parent.Parent.Parent.FullName;
+
+            var state = CheckpointState.LoadCheckpoint(CHECKPOINTPATH);
+
+            using (TrainingSession ts = new TrainingSession(state, TRAININGMODELPATH, EVALMODELPATH, OPTIMIZERMODELPATH))
+            {
+                train(ts, dl, numEpochs, batchSize);
+                Console.WriteLine("############################################################### Training finished");
+
+                var outputNames = new List<string> { "outputs" };
+
+                ts.ExportModelForInferencing(TRAINEDMODELFILE, outputNames);
+            }
+        }
+
+        public static void runInferenceSession(InferenceSession inferenceSession, DataLoader dl, string pathToInference)
+        {
+            Console.WriteLine($"Inferencing now for {pathToInference}");
+
+            var input = dl.imageProcessingForInference(pathToInference);
+            var inputs = new List<NamedOnnxValue> { NamedOnnxValue.CreateFromTensor("pixel_values", input) };
+
+            foreach (var r in inferenceSession.Run(inputs))
+            {
+                var resultsTensor = r.AsTensor<float>();
+
+                printPredictedEmotion(resultsTensor);
+
+                Console.WriteLine(string.Join(",     ", DataLoader.EMOTIONSLABELS));
+                Console.WriteLine(resultsTensor.GetArrayString());
+            }
+        }
+
+        public static float train(TrainingSession ts, DataLoader dl, int numEpochs, int batchSize)
+        {
+            float loss = 0;
+            for (int i = 0; i < numEpochs; i++)
+            {
+                loss = trainEpoch(ts, dl, batchSize);
+
+                Console.WriteLine($"Loss after epoch {i}: {loss}");
+            }
+
+            return loss;
+        }
+
+        public static float trainEpoch(TrainingSession ts, DataLoader dl, int batchSize)
+        {
+            int steps = dl.getNumSteps(batchSize);
+            Console.WriteLine("steps: " + steps);
+
+            float loss = 0;
+
+            for (int step = 0; step < steps; step++)
+            {
+                var inputs = dl.generateBatchInput(batchSize);
+                ts.LazyResetGrad();
+
+                var watch = System.Diagnostics.Stopwatch.StartNew();
+                var outputs = ts.TrainStep(inputs);
+
+                ts.OptimizerStep();
+                watch.Stop();
+                Console.WriteLine($"Execution Time for train step {step} out of {steps}: {watch.ElapsedMilliseconds} ms");
+
+                if(step == steps - 1)
+                {
+                    // if at the last step, update the loss
+                    loss = outputs.ElementAtOrDefault(0).AsTensor<float>().GetValue(0);
+                }
```

Comment:
Shall we accumulate the loss and output the average of the loss across the epoch? Instead of getting the loss associated with the final batch in the epoch?